runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java

```diff
@@ -416,8 +417,13 @@ public FunctionSpec translateDoFn(SdkComponents newComponents) {
 
                 @Override
                 public boolean isStateful() {
-                  return !signature.stateDeclarations().isEmpty()
-                      || !signature.timerDeclarations().isEmpty();
+                  // SDFs should not have user timer/state.
+                  checkState(
+                      signature.stateDeclarations().isEmpty()
+                          && signature.timerDeclarations().isEmpty()
+                          && signature.timerFamilyDeclarations().isEmpty(),
+                      "SDFs should not have user state/timer.");
```

Comment:
This does not seem the right place to prohibit this--we should just pass through what is present. (It is true that some runners may not support it (yet?).)